### PR TITLE
Add quickcache and memoized documentation

### PR DESCRIPTION
**Pull request by esoergel**
This was spawned from a few recent conversations around caching.  It was written in collaboration with @dannyroberts.  (Danny, feel free to push another commit to this branch if you'd like to make amendments).

Pretty view available here: https://github.com/dimagi/commcare-hq/blob/es/cache-docs/docs/caching_and_memoization.rst

### CODE REVIEW

**Comment by esoergel** [comment]
@jmtroth0 @pr33thi @gbova This was written with you all in mind, could you take a look and let me know if you have any questions or if anything is explained poorly?

File: docs/caching_and_memoization.rst [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+``@quickcache`` behaves much more like a normal cache. It stores results in a
+caching backend (Redis, in CCHQ) for a specified timeout (5 minutes, by
+default). This also means they can be shared across worker machines. Quickcache
+also caches objects in local memory (10 seconds, by default). This is faster to
+access than Redis, but its not shared across machines.
+
+Quickcache requires you to specify which arguments to "vary on", that is, which
+arguments uniquely identify a cache
+
+For examples of how it's used, check out `the repo <repo>_`_. For background,
```
Comment (dannyroberts) [inline docs/caching_and_memoization.rst]: I think there's a typo in the link syntax here

**Comment by pr33thi** [comment]
This is awesome! Should the quickcache section also be added to the readme in the quickcache repo?

**Review by dannyroberts — APPROVED** [review]
Looks good, thanks for writing these up! Only had one minor formatting comment

File: docs/caching_and_memoization.rst [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+       def all_objects(self):
+           if self._all_objects is None:
+               result = do_a_bunch_of_stuff()
+               self._all_objects = result
+           return self._all_objects
+
+       def get_object_by_key(self, key):
+           if key not in self._objects_by_key:
+               result = do_a_bunch_of_stuff(key)
+               self._objects_by_key[result]
```
Comment (millerdev) [inline docs/caching_and_memoization.rst]: think this should say `self._objects_by_key[key] = result`

File: docs/caching_and_memoization.rst (hunk 1 above)
Comment (esoergel) [reply docs/caching_and_memoization.rst]: Thanks, fixed

File: docs/caching_and_memoization.rst [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    def _count_users_forms_by_device(self, domain, device_ids):
+        return {
+            FormAccessors(domain.name).count_forms_by_device(device_id)
+            for device_id in device_ids
+        }
+
+    def count_users_forms_by_device(self, domain_obj, user):
+        return self._count_users_forms_by_device(
+            domain_obj.name,
+            [device.device_id for device in user.devices]
```
Comment (millerdev) [inline docs/caching_and_memoization.rst]: will this work with memoized (lists are not hashable)? I think you might need to use something like `frozenset(device.device_id for device in user.devices)`

File: docs/caching_and_memoization.rst [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+Memoized:
+    All arguments must be hashable; notably, lists and dicts are not hashable,
+    but tuples are.
+
+    Return values can be anything.
+
+Quickcache:
+    All vary_on values must be “basic” types (all the way down, if they are
+    collections): string, bool, number, list/tuple (treated as interchangeable),
```
Comment (millerdev) [inline docs/caching_and_memoization.rst]: should "string" be expanded to something like `str` and `unicode` (`bytes` and `str` for Python 3) and they're treated as interchangeable.

File: docs/caching_and_memoization.rst [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+
+Invalidation
+------------
+
+    "There are only two hard problems in computer science - cache invalidation
+    and naming things" (and off-by-one errors)
+
+Memoized doesn’t allow invalidation except by blowing away the whole cache for
+all parameters. Use <function>.reset_cache()
```
Comment (millerdev) [inline docs/caching_and_memoization.rst]: nit: missing code formatting

**Review by millerdev — COMMENTED** [review]
This is great! Few comments.

**Pull request merged**